pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsScheduleOperations failed

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsScheduleOperations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8444903?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8444903?buildTab=artifacts#/) on release-22.1 @ [4f4342ecdf97f478ec8db536643cd48c2f051689](https://github.com/cockroachdb/cockroach/commits/4f4342ecdf97f478ec8db536643cd48c2f051689):


```
=== RUN   TestSQLStatsScheduleOperations
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/f6543d871e87ca5f79c9e9854efd0ecf/logTestSQLStatsScheduleOperations2634254611
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestSQLStatsScheduleOperations
    scheduled_sql_stats_compaction_test.go:270: -- test log scope end --
--- FAIL: TestSQLStatsScheduleOperations (47.00s)
=== RUN   TestSQLStatsScheduleOperations/warn_schedule_long_run_interval/via_cluster_setting
    scheduled_sql_stats_compaction_test.go:236: 
        	Error Trace:	sql_runner.go:74
        	            				sql_runner.go:235
        	            				scheduled_sql_stats_compaction_test.go:236
        	Error:      	Received unexpected error:
        	            	query '
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:238
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:212
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:74
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:235
        	            	  | pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test.TestSQLStatsScheduleOperations.func3.1
        	            	  | 	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/scheduled_sql_stats_compaction_test.go:236
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1259
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) query '
        	            	  | SELECT schedule_expr
        	            	  | FROM system.scheduled_jobs WHERE schedule_id = 834207426563211265': expected:
        	            	  | @hourly
        	            	  | got:
        	            	  | 59 23 24 2 ?
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestSQLStatsScheduleOperations/warn_schedule_long_run_interval/via_cluster_setting
        --- FAIL: TestSQLStatsScheduleOperations/warn_schedule_long_run_interval/via_cluster_setting (45.96s)
=== RUN   TestSQLStatsScheduleOperations/warn_schedule_long_run_interval
    --- FAIL: TestSQLStatsScheduleOperations/warn_schedule_long_run_interval (46.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsScheduleOperations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23763

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsScheduleOperations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8490786?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8490786?buildTab=artifacts#/) on release-22.1 @ [97d9ad702b767a81260cffba646f58b5dd34b69e](https://github.com/cockroachdb/cockroach/commits/97d9ad702b767a81260cffba646f58b5dd34b69e):


```
=== RUN   TestSQLStatsScheduleOperations
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/f6543d871e87ca5f79c9e9854efd0ecf/logTestSQLStatsScheduleOperations4119994655
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestSQLStatsScheduleOperations
    scheduled_sql_stats_compaction_test.go:270: -- test log scope end --
--- FAIL: TestSQLStatsScheduleOperations (47.20s)
=== RUN   TestSQLStatsScheduleOperations/warn_schedule_long_run_interval/via_cluster_setting
    scheduled_sql_stats_compaction_test.go:236: 
        	Error Trace:	sql_runner.go:74
        	            				sql_runner.go:235
        	            				scheduled_sql_stats_compaction_test.go:236
        	Error:      	Received unexpected error:
        	            	query '
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:238
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:212
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:74
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:235
        	            	  | pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test.TestSQLStatsScheduleOperations.func3.1
        	            	  | 	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/scheduled_sql_stats_compaction_test.go:236
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1259
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) query '
        	            	  | SELECT schedule_expr
        	            	  | FROM system.scheduled_jobs WHERE schedule_id = 835059096103747585': expected:
        	            	  | @hourly
        	            	  | got:
        	            	  | 59 23 24 2 ?
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestSQLStatsScheduleOperations/warn_schedule_long_run_interval/via_cluster_setting
        --- FAIL: TestSQLStatsScheduleOperations/warn_schedule_long_run_interval/via_cluster_setting (46.05s)
=== RUN   TestSQLStatsScheduleOperations/warn_schedule_long_run_interval
    --- FAIL: TestSQLStatsScheduleOperations/warn_schedule_long_run_interval (46.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsScheduleOperations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/sqlstats/persistedsqlstats/scheduled_job_monitor.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/scheduled_job_monitor.go
@@ -99,10 +99,14 @@ func (j *jobMonitor) start(ctx context.Context, stopper *stop.Stopper) {
 			case <-stopCtx.Done():
 				return
 			}
-			if SQLStatsCleanupRecurrence.Get(&j.st.SV) != currentRecurrence || nextJobScheduleCheck.Before(timeutil.Now()) {
-				j.updateSchedule(stopCtx, j.ie)
+
+			// Read the config once to avoid race condition if config is changed during
+			// the update process.
+			newRecurrence := SQLStatsCleanupRecurrence.Get(&j.st.SV)
+			if newRecurrence != currentRecurrence || nextJobScheduleCheck.Before(timeutil.Now()) {
+				j.updateSchedule(stopCtx, j.ie, newRecurrence)
 				nextJobScheduleCheck = timeutil.Now().Add(j.jitterFn(j.scanInterval))
-				currentRecurrence = SQLStatsCleanupRecurrence.Get(&j.st.SV)
+				currentRecurrence = newRecurrence
 			}
 
 			timer.Reset(updateCheckInterval)
@@ -141,7 +145,9 @@ func (j *jobMonitor) getSchedule(
 	return sj, nil
 }
 
-func (j *jobMonitor) updateSchedule(ctx context.Context, ie sqlutil.InternalExecutor) {
+func (j *jobMonitor) updateSchedule(
+	ctx context.Context, ie sqlutil.InternalExecutor, cronExpr string,
+) {
 	var sj *jobs.ScheduledJob
 	var err error
 	retryOptions := retry.Options{
@@ -162,8 +168,7 @@ func (j *jobMonitor) updateSchedule(ctx context.Context, ie sqlutil.InternalExec
 					return err
 				}
 			}
-			// Update schedule with new recurrence, if different.
-			cronExpr := SQLStatsCleanupRecurrence.Get(&j.st.SV)
+
 			if sj.ScheduleExpr() == cronExpr {
 				return nil
 			}

--- a/pkg/sql/sqlstats/persistedsqlstats/scheduled_sql_stats_compaction_test.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/scheduled_sql_stats_compaction_test.go
@@ -234,6 +234,11 @@ func TestSQLStatsScheduleOperations(t *testing.T) {
 
 			helper.sqlDB.Exec(t, "RESET CLUSTER SETTING sql.stats.cleanup.recurrence")
 			helper.sqlDB.CheckQueryResultsRetry(t,
+				`SHOW CLUSTER SETTING sql.stats.cleanup.recurrence`,
+				[][]string{{"@hourly"}},
+			)
+
+			helper.sqlDB.CheckQueryResultsRetry(t,
 				fmt.Sprintf(`
 SELECT schedule_expr
 FROM system.scheduled_jobs WHERE schedule_id = %d`, schedID),

